### ISSUE TITLE
Lack of folder structure

### ISSUE THREAD

**Issue by Silverwolf90** [0xfe/vexflow#403]
VexFlow currently has 75 items sitting in `/src` and could benefit from a little more structure. What if we move to a folder structure that more closely resembles the [class hierarchy](https://github.com/0xfe/vexflow/wiki/VexFlow-Class-Hierarchy)?

**Comment by aaronmars** [comment]
Hmmm... I see what you're saying, but from a usage perspective, I feel the flat structure is a bit easier.

If I wanted to use VexFlow unbundled, I'd have things like `import {StaveNote} from 'src/StaveNote.js';` in my code. Knowing the entire library is in one place makes this use case easier.

Of course, the bundled library hides all of this by packaging it all up under the `Vex` global, and the folder structure doesn't matter.

A middle ground might be to identify the public API and put those modules into a "lib" directory or something.  Then, all of the internal/helper things could be structured however is best.

**Comment by Silverwolf90** [comment]
To use VexFlow unbundled, couldn't one just do this?

```javascript
import Vex from 'src';
const { StaveNote } = Vex.Flow;
```

Or in `src/index.js` we could `export` each module (while still attaching them to the `Vex.Flow`). Then you could do:

```javascript
import { StaveNote } from './src'
```

The biggest problem about the lack of folders is that it provides no indication to how the codebase is structured. Just looking at `/src` does not provide any high-level understanding of a codebase -- things that are alike are not necessarily close in proximity. This has a negative impact on codebase accessibility for new contributors. I think that's more important than making the unbundled use case slightly more difficult.

**Comment by aaronmars** [comment]
OK.  I'm actually talking about bypassing the Vex global and the bundle represented in index.js.  Rather than bring in the kitchen sink, it's possible to just piecemeal the VF bits that I want, and let the dependency tree happen.  For example, if my app doesn't use the tablature parts of VexFlow, why bring it in?

So, a flat directory structure isn't so bad to me since the dependency graph defines the structure of the code.  My concern is the paths to the modules would become complicated.

Having said all of that, I can see the other side of it in that the import paths might be a bit more self-documenting.  Also, to your point, it will make things much easier for those of us that develop in the library.  I guess I'm just pointing out that this would be a breaking change for those of us that want to import the ES6 sources directly, before they are transpiled and bundled (I'm pretty sure it's not a problem at this point, but it will be in the future once the need for transpilation goes away).

**Comment by Silverwolf90** [comment]
You raise some good points. The fact that VexFlow brings in all the dependencies regardless of what is specifically needed is problematic. We could benefit a lot from the ability to create custom builds. Given VexFlow's structure, that shouldn't be hard to do. Most people don't need the entire codebase. In fact, I'd bet most people only use a very small subset of the codebase.

Nonetheless, easily enabling piecemeal builds is not the burden of the file structure. That's on us to build better tooling around the codebase (which shouldn't be that hard... I think?)

Out of curiosity, is this use-case how you're using VF currently? Or just something you're throwing out in general?

(I'd also like to point out that this usecase precariously relies on us not using the global `Vex` object to access anything that is attached to it in `src/index.js`)

**Comment by gristow** [comment]
I currently do this -- a custom build minus tablature, Raphael & canvas, etc... I'd love to see us let the dependency graph generate includes of the build. (Which would also let me easily pull in just the code I need via imports in my own Webpack build process.)

**Comment by Silverwolf90** [comment]
Ok cool, I'll look into this. With a a little code wrapped around webpack I
feel like customizable builds should be certainly possible.

**Comment by aaronmars** [comment]
Yeah.  I have tested it, and it works great.  I basically made a project, and instead of just loading vexflow-min.js, I imported the VF modules directly where they are being used.  Then, I Rollup/webpack *my app*, and that process sucks in the VF modules where needed.

Of course, this doesn't mean the bundled, kitchen sink thing isn't important, it's just that we have an alternative to it, and it's already available for free*.

*As long as the developer doing it understands the architecture and is willing and able to do the bundling at the app level.
